pkg/controllers/nodetemplate/controller.go

```diff
@@ -30,48 +29,60 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	"github.com/aws/aws-sdk-go/service/ec2"
-	"github.com/samber/lo"
-
 	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter-core/pkg/utils/result"
 	"github.com/aws/karpenter/pkg/apis/v1alpha1"
 	"github.com/aws/karpenter/pkg/providers/amifamily"
 	"github.com/aws/karpenter/pkg/providers/securitygroup"
 	"github.com/aws/karpenter/pkg/providers/subnet"
 )
 
+type nodeTemplateReconciler interface {
+	Reconcile(context.Context, *v1alpha1.AWSNodeTemplate) (reconcile.Result, error)
+}
+
 var _ corecontroller.TypedController[*v1alpha1.AWSNodeTemplate] = (*Controller)(nil)
 
 type Controller struct {
-	kubeClient            client.Client
-	subnetProvider        *subnet.Provider
-	securityGroupProvider *securitygroup.Provider
-	amiProvider           *amifamily.Provider
+	kubeClient client.Client
+	status     *Status
+	hash       *Hash
 }
 
-func NewController(kubeClient client.Client, subnetProvider *subnet.Provider, securityGroups *securitygroup.Provider, amiprovider *amifamily.Provider) corecontroller.Controller {
+func NewController(kubeClient client.Client, subnetProvider *subnet.Provider, securityGroupProvider *securitygroup.Provider, amiProvider *amifamily.Provider) corecontroller.Controller {
 	return corecontroller.Typed[*v1alpha1.AWSNodeTemplate](kubeClient, &Controller{
-		kubeClient:            kubeClient,
-		subnetProvider:        subnetProvider,
-		securityGroupProvider: securityGroups,
-		amiProvider:           amiprovider,
+		kubeClient: kubeClient,
+		status:     &Status{subnetProvider: subnetProvider, securityGroupProvider: securityGroupProvider, amiProvider: amiProvider},
+		hash:       &Hash{},
 	})
 }
 
 func (c *Controller) Reconcile(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate) (reconcile.Result, error) {
 	stored := nodeTemplate.DeepCopy()
 
-	err := multierr.Combine(
-		c.resolveSubnets(ctx, nodeTemplate),
-		c.resolveSecurityGroups(ctx, nodeTemplate),
-		c.resolveAMIs(ctx, nodeTemplate),
-	)
+	var results []reconcile.Result
+	var errs error
+	reconcilers := []nodeTemplateReconciler{
+		c.status,
+		c.hash,
```

Comment:
likewise for my other comment, this has no chance to fail, so I think you may be over-abstracting this. Personally, I think you should be able to leave it as it was and just always idempotently add in the hash annotation.